examples/api-samples/src/browser/api-samples-frontend-module.ts

```diff
@@ -42,4 +47,176 @@ export default new ContainerModule((
     bindSampleFilteredCommandContribution(bind);
     bindSampleToolbarContribution(bind, rebind);
     bindMonacoPreferenceExtractor(bind);
+    bindSampleStartupTask(bind, rebind);
 });
+
+function bindSampleStartupTask(bind: interfaces.Bind, rebind: interfaces.Rebind): void {
```

Comment:
Move everything below here into its own file.

Reply:
> I'm perfectly happy with the changes to the WorkspacesService: I don't see any reason that information shouldn't be available to those methods.

Thank you so much for the self-assignment.

> I have some doubts about the sample API contribution, though.

I wanted to drop the example before the merge. Let me know if the API changes are OK, and I will remove the last commit.

The sole purpose of the sample was to show that custom behavior can be achieved by overriding the `openNewWindow` and `reloadWindow` with the proposed API changes. In the Arduino IDE, we would like to make the Theia version updates easy and avoid unnecessary customizations, such as [`doOpen`](https://github.com/arduino/arduino-ide/blob/4907ef2a478034d0b9f9436e15988a1833768c3d/arduino-ide-extension/src/browser/theia/workspace/workspace-service.ts#L132) and [`openWindow`](https://github.com/arduino/arduino-ide/blob/4907ef2a478034d0b9f9436e15988a1833768c3d/arduino-ide-extension/src/browser/theia/workspace/workspace-service.ts#L146-L149).

> My impression of the new feature is that it's basically intended to be a CLI of sorts, but for the browser: a way to pass arguments that can be interpreted as the downstream application wishes. 

I appreciate that you have taken the time to think this further. It is precisely the bahevior we want, but it's not yet clear how to support it properly in the long run. URLs have a limited length, so URL manipulation is not the ultimate solution.

> Until now, we've had one 'positional' argument, the path to the workspace, but this PR envisions (reasonably) that we might want more. If we want to go this route, might it make sense to build some infrastructure around it - the code that clears the additional arguments from the URL makes me a little nervous.

Absolutely. When we have a distilled approach, I am happy to open a GH discussion or PR and decide on the details with the core committers. Currently, IDE2's top priority is to run in electron. Theia requires a browser and electron solution.

> Another question is whether this is the best way to accomplish the objective. The use-case seems to be cases where one window is opened from another window, since anything else could be handled by the actual CLI. If that's the use case, then rather than passing arguments in the URL, inter-window messaging could be used, since the parent window has access to the child window and could send it a message encoding the details of its 'startup task' when the window was ready to handle it. That would lose some of the flexibility of the URL-based approach - it could never be used for the first window - but it would avoid messing with the URL at all and so confusing the existing workspace setup.

Great idea. I also thought about it but decided to use the URL workaround for now. Here is why:
 - Window A opens B and wants to send an "initial config" via IPC to window B,
 - While B is loading (`did-finish-load` not fired yet), window A closes,
 - B might never receive the message.

Please correct me if I am conceptually wrong or if you have a better idea. Thank you!

Reply:
> I wanted to drop the example before the merge. Let me know if the API changes are OK, and I will remove the last commit.

Sounds good - if you drop that commit, I'll happily approve 👍 .